Mathlib/Analysis/Complex/Hadamard.lean

```diff
@@ -0,0 +1,488 @@
+          simp only [norm_eq_abs, abs_cpow_eq_rpow_re_of_pos h0, sub_re, one_re,
+          Real.rpow_le_one_of_one_le_of_nonpos hM0_one (sub_nonpos.mpr hset.2)]
+        -- `0 < (sSupAbsIm f 0) < 1`
+        · rw [not_le] at hM0_one; apply le_trans _ (le_max_right _ _)
+          simp only [norm_eq_abs, abs_cpow_eq_rpow_re_of_pos h0, sub_re, one_re]
+          apply Real.rpow_le_rpow_of_exponent_ge (h0) (le_of_lt hM0_one) _
+          simp only [neg_le_sub_iff_le_add, le_add_iff_nonneg_left, hset.1]
+      · by_cases hM1_one : 1 ≤ sSupAbsIm f 1
+        -- `1 ≤ sSupAbsIm f 1`
+        · apply le_trans _ (le_max_left _ _)
+          simp only [norm_eq_abs, abs_cpow_eq_rpow_re_of_pos h1, sub_re, one_re, neg_re,
+          Real.rpow_le_one_of_one_le_of_nonpos hM1_one (Right.neg_nonpos_iff.mpr hset.1)]
+        -- `0 < sSupAbsIm f 1 < 1`
+        · rw [not_le] at hM1_one; apply le_trans _ (le_max_right _ _)
+          simp only [norm_eq_abs, abs_cpow_eq_rpow_re_of_pos h1, sub_re, one_re, neg_re,
+          Real.rpow_le_rpow_of_exponent_ge (h1) (le_of_lt hM1_one)
+            (neg_le_neg_iff.mpr hset.2)]
+    simp only [gt_iff_lt, lt_max_iff, zero_lt_one, true_or, mul_nonneg_iff_of_pos_left, le_max_iff,
+      zero_le_one]
+
+-- Proof that the edges are bounded by one
+lemma F_edge_le_one (hB : BddAbove ((norm ∘ f) '' (closedStrip 0 1)))
+    (hz : z ∈ re ⁻¹' {0, 1}) : ‖F f z‖ ≤ 1 := by
+  simp only [F, invInterpStrip, norm_smul, norm_eq_abs, map_mul, abs_cpow_eq_rpow_re_of_pos,
+    h0, h1, sub_re, one_re, neg_re]
+  cases' hz with hz0 hz1
+  -- `z.re = 0`
+  · simp [hz0, mul_one, zero_sub, Real.rpow_zero, neg_zero,
+    Real.rpow_neg_one, inv_mul_le_iff h0]
+    rw [← hz0]
+    apply abs_le_sSupAbsIm f z _ hB
+    simp only [closedStrip, mem_preimage, zero_le_one, left_mem_Icc, hz0]
+  -- `z.re = 1`
+  · rw [mem_singleton_iff] at hz1
+    simp only [hz1, one_mul, Real.rpow_zero, sub_self, Real.rpow_neg_one,
+    inv_mul_le_iff h1, mul_one]
+    rw [← hz1]
+    apply abs_le_sSupAbsIm f z _ hB
+    simp only [closedStrip, mem_preimage, zero_le_one, hz1, right_mem_Icc]
+
+theorem abs_mul_invInterpStrip_le_one_on_closedStrip
+    (hd : DiffContOnCl ℂ f (strip 0 1)) (hB : BddAbove ((norm ∘ f) '' (closedStrip 0 1)))
+    (z : ℂ) (hz : z ∈ closedStrip 0 1) : ‖F f z‖ ≤ 1 := by
+  apply PhragmenLindelof.vertical_strip
+    (DiffContOnCl.smul (diffContOnCl_invInterpStrip f h0 h1) hd) _
+    (fun x hx ↦ F_edge_le_one x f h0 h1 hB (Or.inl hx))
+    (fun x hx ↦ F_edge_le_one x f h0 h1 hB (Or.inr hx)) hz.1 hz.2
+  use 0
+  rw [sub_zero, div_one]
+  refine ⟨ Real.pi_pos, ?_⟩
+  obtain ⟨BF, hBF⟩ := F_BddAbove f h0 h1 hB
+  simp [mem_upperBounds] at hBF
+  use BF
+  rw [Asymptotics.isBigO_iff]
+  use 1
+  rw [eventually_inf_principal]
+  apply eventually_of_forall
+  intro x hx
+  norm_num
+  apply le_trans (hBF x ((preimage_mono Ioo_subset_Icc_self) hx))
+    (le_trans (le_of_lt (lt_add_one BF)) (Real.add_one_le_exp BF))
```

Comment:
```suggestion
  exact (hBF x ((preimage_mono Ioo_subset_Icc_self) hx)).trans
    ((lt_add_one BF).le.trans (Real.add_one_le_exp BF))
```
Typically one uses `exact` instead of `apply` when closing a proof. Also, dot notation is very convenient for chains of inequalities.

Reply:
You're right that's pretty cool!